libopae/plugins/xfpga/metrics/threshold.c

```diff
@@ -0,0 +1,469 @@
+#endif // HAVE_CONFIG_H
+
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <string.h>
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <glob.h>
+
+#include "types_int.h"
+#include "safe_string/safe_string.h"
+#include "metrics_int.h"
+#include "common_int.h"
+#include "metrics/bmc/bmc.h"
+#include "metrics_int.h"
+#include "metrics_max10.h"
+#include "threshold.h"
+
+
+fpga_result xfpga_fpgaGetThresholdInfo(fpga_handle handle,
+					metric_threshold *metric_thresholds,
+					uint32_t *num_thresholds)
+{
+	fpga_result result             = FPGA_OK;
+	struct _fpga_token *_token     = NULL;
+	enum fpga_hw_type hw_type     = FPGA_HW_UNKNOWN;
+	fpga_objtype objtype;
+
+
+	if (handle == NULL ||
+		(metric_thresholds == NULL &&
+		num_thresholds == NULL)) {
+		FPGA_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	struct _fpga_handle *_handle = (struct _fpga_handle *)handle;
+	if (_handle == NULL) {
+		FPGA_ERR("Invalid handle");
+		return FPGA_INVALID_PARAM;
+	}
+
+	_token = (struct _fpga_token *)_handle->token;
+	if (_token == NULL) {
+		FPGA_ERR("Invalid token within handle");
+		return FPGA_INVALID_PARAM;
+	}
+
+	result = get_fpga_object_type(handle, &objtype);
+	if (result != FPGA_OK) {
+		FPGA_ERR("Failed to object type");
+		return result;
+	}
+
+	if (objtype != FPGA_DEVICE) {
+		FPGA_ERR("FPGA object type is not FPGA DEVICE ");
+		return result;
+	}
+
+	// get fpga hw type.
+	result = get_fpga_hw_type(_handle, &hw_type);
+	if (result != FPGA_OK) {
+		FPGA_ERR("Failed to discover hardware type.");
+		return result;
+	}
+
+	switch (hw_type) {
+	// MCP
+	case FPGA_HW_MCP: {
+		FPGA_ERR("Not Supported MCP thresholds.");
+		result = FPGA_EXCEPTION;
+	}
+	break;
+
+	// DCP RC
+	case FPGA_HW_DCP_RC: {
+
+		result = get_bmc_threshold_info(handle,
+			metric_thresholds, num_thresholds);
+		if (result != FPGA_OK) {
+			FPGA_ERR("Failed to get bmc thresholds.");
+			return result;
+		}
+
+	}
+	break;
+
+	// VC DC
+	case FPGA_HW_DCP_DC:
+	case FPGA_HW_DCP_VC: {
+		// Max10
+		result = get_max10_threshold_info(handle,
+			metric_thresholds, num_thresholds);
+		if (result != FPGA_OK) {
+			FPGA_ERR("Failed to get max10 thresholds.");
+			return result;
+		}
+
+	}
+	break;
+
+	default:
+		FPGA_ERR("Unknown Device ID.");
+		result = FPGA_EXCEPTION;
+	}
+
+	return result;
+}
+
+fpga_result get_bmc_threshold_info(fpga_handle handle,
+					metric_threshold *metric_thresholds,
+					uint32_t *num_thresholds)
+{
+	fpga_result result                = FPGA_OK;
+	uint32_t num_sensors              = 0;
+	uint32_t num_values               = 0;
+	uint32_t x                        = 0;
+	errno_t e                         = 0;
+
+	sdr_details details;
+	bmc_sdr_handle records;
+	bmc_values_handle values;
+
+	if (handle == NULL ||
+		num_thresholds == NULL) {
+		FPGA_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	struct _fpga_handle *_handle = (struct _fpga_handle *)handle;
+
+	if (_handle->bmc_handle == NULL)
```

Comment:
Need to lock the _handle around this check and the subsequent assignment of _handle->bmc_handle.

Reply:
ie
```c
lock(&_handle->lock);
if (_handle->bmc_handle == NULL)
   _handle->bmc_handle = ...
unlock(&_handle->lock);
```